File: 算法/Python/第3章 排序算法/3.4 插入排序/insertion_sort.py
def insertion_sort(list):
    """
        插入排序
        Args:
            list (list): 待排序列表
    """
    n = len(list)
    for i in range(1, n):
        temp = list[i]
        j = i - 1
        while j >= 0 and temp < list[j]:
            list[j+1] = list[j]
            j -= 1
        list[j+1] = temp

def binary_insertion_sort(lst):
    n = len(lst)
    for i in range(1, n):
        temp = lst[i]
        start = 0
        end = i - 1
        while start <= end:
            mid = start + (end - start) // 2
            if lst[mid] > temp:
                end = mid - 1
            else:
                start = mid + 1
        j = i - 1
        while j > end:
            lst[j+1] = lst[j]
            j -= 1
        lst[j+1] = temp

def main():
    list = [5, 8, 6, 3, 9, 2, 1, 7]
    print("排序前：" + str(list))
    
    # 未优化
    insertion_sort(list)
    # 优化
    # binary_insertion_sort(list)
    
    print("排序后：" + str(list))

if __name__ == "__main__":
    main()

File: 算法/C/第3章 排序算法/3.4 插入排序/insertion_sort.c
#include <stdio.h>

/**
 * @brief  打印数组
 * @param  arr: 数组
 * @param  n: 数组长度
 * @retval None
 */
void showArr(int *arr, int n) {
    for(int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

/**
 * @brief  插入排序
 * @param  arr: 数组
 * @param  n: 数组长度
 */
void insertionSort(int *arr, int n) {
    for(int i = 1; i < n; i++) {
        int temp = arr[i];
        int j = i - 1;
        while(j >= 0 && temp < arr[j]) {
            arr[j+1] = arr[j];
            j--;
        }
        arr[j+1] = temp;
    }
}

void binaryInsertionSort(int *arr, int n) {
    for(int i = 1; i < n; i++) {
        int temp = arr[i];
        int start = 0;
        int end = i - 1;
        while(start <= end) {
            int mid = start + (end - start) / 2;
            if(arr[mid] > temp) {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }
        int j;
        for(j = i - 1; j > end; j--) {
            arr[j+1] = arr[j];
        }
        arr[j+1] = temp;
    }
}

int main() {
    int arr[] = {5, 8, 6, 3, 9, 2, 1, 7};
    int n = sizeof(arr) / sizeof(arr[0]);
    
    printf("排序前：");
    showArr(arr, n);
	
	// 未优化
    insertionSort(arr, n);
	// 优化
	// binaryInsertionSort(arr, n);
	
    printf("排序后：");
    showArr(arr, n);
    return 0;
}